// @jbrowse/plugin-alignments/dist/LinearSNPCoverageDisplay/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const configSchema_1 = __importDefault(require("./models/configSchema"));
const model_1 = __importDefault(require("./models/model"));
const plugin_wiggle_1 = require("@jbrowse/plugin-wiggle");
const DisplayType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/DisplayType"));
function register(pluginManager) {
    pluginManager.addDisplayType(() => {
        const configSchema = (0, configSchema_1.default)(pluginManager);
        return new DisplayType_1.default({
            name: 'LinearSNPCoverageDisplay',
            configSchema,
            stateModel: (0, model_1.default)(pluginManager, configSchema),
            trackType: 'AlignmentsTrack',
            viewType: 'LinearGenomeView',
            ReactComponent: plugin_wiggle_1.LinearWiggleDisplayReactComponent,
        });
    });
}
exports.default = register;
