// @jbrowse/plugin-alignments/dist/LinearAlignmentsDisplay/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const DisplayType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/DisplayType"));
// locals
const configSchema_1 = __importDefault(require("./models/configSchema"));
const model_1 = __importDefault(require("./models/model"));
const AlignmentsDisplay_1 = __importDefault(require("./components/AlignmentsDisplay"));
function default_1(pluginManager) {
    pluginManager.addDisplayType(() => {
        const configSchema = (0, configSchema_1.default)(pluginManager);
        return new DisplayType_1.default({
            name: 'LinearAlignmentsDisplay',
            configSchema,
            stateModel: (0, model_1.default)(pluginManager, configSchema),
            trackType: 'AlignmentsTrack',
            viewType: 'LinearGenomeView',
            ReactComponent: AlignmentsDisplay_1.default,
        });
    });
}
exports.default = default_1;
