// @jbrowse/plugin-authentication/dist/GoogleDriveOAuthModel/configSchema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const configuration_1 = require("@jbrowse/core/configuration");
const configSchema_1 = __importDefault(require("../OAuthModel/configSchema"));
/**
 * #config GoogleDriveOAuthInternetAccount
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const GoogleDriveOAuthConfigSchema = (0, configuration_1.ConfigurationSchema)('GoogleDriveOAuthInternetAccount', {
    /**
     * #slot
     */
    authEndpoint: {
        description: 'the authorization code endpoint of the internet account',
        type: 'string',
        defaultValue: 'https://accounts.google.com/o/oauth2/v2/auth',
    },
    /**
     * #slot
     */
    scopes: {
        description: 'optional scopes for the authorization call',
        type: 'string',
        defaultValue: 'https://www.googleapis.com/auth/drive.readonly',
    },
    /**
     * #slot
     */
    domains: {
        description: 'array of valid domains the url can contain to use this account',
        type: 'stringArray',
        defaultValue: ['drive.google.com"'],
    },
    /**
     * #slot
     */
    responseType: {
        description: 'the type of response from the authorization endpoint',
        type: 'string',
        defaultValue: 'token',
    },
}, {
    /**
     * #baseConfiguration
     */
    baseConfiguration: configSchema_1.default,
    explicitlyTyped: true,
});
exports.default = GoogleDriveOAuthConfigSchema;
