// @jbrowse/plugin-wiggle/dist/LinearWiggleDisplay/components/SetColorDialog.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const mobx_react_1 = require("mobx-react");
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const Close_1 = __importDefault(require("@mui/icons-material/Close"));
const ColorPicker_1 = require("@jbrowse/core/ui/ColorPicker");
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    closeButton: {
        position: 'absolute',
        right: theme.spacing(1),
        top: theme.spacing(1),
        color: theme.palette.grey[500],
    },
}));
function SetColorDialog({ model, handleClose, }) {
    const { classes } = useStyles();
    const [posneg, setPosNeg] = (0, react_1.useState)(false);
    return (react_1.default.createElement(material_1.Dialog, { open: true, onClose: handleClose },
        react_1.default.createElement(material_1.DialogTitle, null,
            "Select either an overall color, or the positive/negative colors. Note that density renderers only work properly with positive/negative colors",
            react_1.default.createElement(material_1.IconButton, { "aria-label": "close", className: classes.closeButton, onClick: handleClose },
                react_1.default.createElement(Close_1.default, null))),
        react_1.default.createElement(material_1.DialogContent, null,
            react_1.default.createElement(material_1.FormControlLabel, { checked: !posneg, onClick: () => setPosNeg(false), control: react_1.default.createElement(material_1.Radio, null), label: 'Overall color' }),
            react_1.default.createElement(material_1.FormControlLabel, { checked: posneg, onClick: () => setPosNeg(true), control: react_1.default.createElement(material_1.Radio, null), label: 'Positive/negative color' }),
            posneg ? (react_1.default.createElement(react_1.default.Fragment, null,
                react_1.default.createElement(material_1.Typography, null, "Positive color"),
                react_1.default.createElement(ColorPicker_1.ColorPicker, { color: model.posColor || 'black', onChange: event => {
                        model.setPosColor(event);
                        model.setColor(undefined);
                    } }),
                react_1.default.createElement(material_1.Typography, null, "Negative color"),
                react_1.default.createElement(ColorPicker_1.ColorPicker, { color: model.negColor || 'black', onChange: event => {
                        model.setNegColor(event);
                        model.setColor(undefined);
                    } }))) : (react_1.default.createElement(react_1.default.Fragment, null,
                react_1.default.createElement(material_1.Typography, null, "Overall color"),
                react_1.default.createElement(ColorPicker_1.ColorPicker, { color: model.color || 'black', onChange: event => model.setColor(event) })))),
        react_1.default.createElement(material_1.DialogActions, null,
            react_1.default.createElement(material_1.Button, { onClick: () => {
                    model.setPosColor(undefined);
                    model.setNegColor(undefined);
                    model.setColor(undefined);
                }, color: "secondary", variant: "contained" }, "Restore default"),
            react_1.default.createElement(material_1.Button, { variant: "contained", color: "primary", type: "submit", onClick: () => handleClose() }, "Submit"))));
}
exports.default = (0, mobx_react_1.observer)(SetColorDialog);
