// @jbrowse/plugin-sequence/dist/TwoBitAdapter/TwoBitAdapter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const BaseAdapter_1 = require("@jbrowse/core/data_adapters/BaseAdapter");
const io_1 = require("@jbrowse/core/util/io");
const rxjs_1 = require("@jbrowse/core/util/rxjs");
const simpleFeature_1 = __importDefault(require("@jbrowse/core/util/simpleFeature"));
const twobit_1 = require("@gmod/twobit");
const configuration_1 = require("@jbrowse/core/configuration");
class TwoBitAdapter extends BaseAdapter_1.BaseSequenceAdapter {
    async initChromSizes() {
        const conf = (0, configuration_1.readConfObject)(this.config, 'chromSizesLocation');
        // check against default and empty in case someone makes the field blank in
        // config editor, may want better way to check "optional config slots" in
        // future
        if (conf.uri !== '/path/to/default.chrom.sizes' && conf.uri !== '') {
            const file = (0, io_1.openLocation)(conf, this.pluginManager);
            const data = await file.readFile('utf8');
            return Object.fromEntries(data === null || data === void 0 ? void 0 : data.split(/\n|\r\n|\r/).filter(line => !!line.trim()).map(line => {
                const [name, length] = line.split('\t');
                return [name, +length];
            }));
        }
        return undefined;
    }
    constructor(config, getSubAdapter, pluginManager) {
        super(config, getSubAdapter, pluginManager);
        this.chromSizesData = this.initChromSizes();
        this.twobit = new twobit_1.TwoBitFile({
            filehandle: (0, io_1.openLocation)((0, configuration_1.readConfObject)(config, 'twoBitLocation'), this.pluginManager),
        });
    }
    async getRefNames() {
        const chromSizesData = await this.chromSizesData;
        if (chromSizesData) {
            return Object.keys(chromSizesData);
        }
        return this.twobit.getSequenceNames();
    }
    async getRegions() {
        const chromSizesData = await this.chromSizesData;
        if (chromSizesData) {
            return Object.keys(chromSizesData).map(refName => ({
                refName,
                start: 0,
                end: chromSizesData[refName],
            }));
        }
        const refSizes = await this.twobit.getSequenceSizes();
        return Object.keys(refSizes).map(refName => ({
            refName,
            start: 0,
            end: refSizes[refName],
        }));
    }
    /**
     * Fetch features for a certain region
     * @param param -
     * @returns Observable of Feature objects in the region
     */
    getFeatures({ refName, start, end }) {
        return (0, rxjs_1.ObservableCreate)(async (observer) => {
            const size = await this.twobit.getSequenceSize(refName);
            const regionEnd = size !== undefined ? Math.min(size, end) : end;
            const seq = await this.twobit.getSequence(refName, start, regionEnd);
            if (seq) {
                observer.next(new simpleFeature_1.default({
                    id: `${refName} ${start}-${regionEnd}`,
                    data: { refName, start, end: regionEnd, seq },
                }));
            }
            observer.complete();
        });
    }
    /**
     * called to provide a hint that data tied to a certain region
     * will not be needed for the forseeable future and can be purged
     * from caches, etc
     */
    freeResources( /* { region } */) { }
}
exports.default = TwoBitAdapter;
