// @jbrowse/plugin-legacy-jbrowse/dist/JBrowse1Connection/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const models_1 = require("@jbrowse/core/pluggableElementTypes/models");
const configuration_1 = require("@jbrowse/core/configuration");
/**
 * #config JBrowse1Connection
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const JBrowse1Connection = (0, configuration_1.ConfigurationSchema)('JBrowse1Connection', {
    /**
     * #slot
     */
    dataDirLocation: {
        type: 'fileLocation',
        defaultValue: {
            uri: 'http://mysite.com/jbrowse/data/',
            locationType: 'UriLocation',
        },
        description: 'the location of the JBrowse 1 data directory, often something like http://mysite.com/jbrowse/data/',
    },
    /**
     * #slot
     */
    assemblyNames: {
        description: 'name of the assembly the connection belongs to, should be a single entry',
        type: 'stringArray',
        defaultValue: [],
    },
}, {
    /**
     * #baseConfiguration
     */
    baseConfiguration: models_1.baseConnectionConfig,
});
exports.default = JBrowse1Connection;
