// @jbrowse/plugin-alignments/dist/LinearAlignmentsDisplay/components/AlignmentsDisplay.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const mobx_react_1 = require("mobx-react");
const mui_1 = require("tss-react/mui");
const configuration_1 = require("@jbrowse/core/configuration");
const ui_1 = require("@jbrowse/core/ui");
const useStyles = (0, mui_1.makeStyles)()(() => ({
    resizeHandle: {
        height: 2,
        position: 'absolute',
        zIndex: 2,
    },
}));
function AlignmentsDisplay({ model }) {
    const { PileupDisplay, SNPCoverageDisplay, showPileup, showCoverage } = model;
    const { classes } = useStyles();
    const top = SNPCoverageDisplay.height;
    return (react_1.default.createElement("div", { "data-testid": `display-${(0, configuration_1.getConf)(model, 'displayId')}`, style: { position: 'relative' } },
        showCoverage ? (react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement("div", { "data-testid": "Blockset-snpcoverage" },
                react_1.default.createElement(SNPCoverageDisplay.RenderingComponent, { model: SNPCoverageDisplay })),
            react_1.default.createElement(ui_1.ResizeHandle, { onDrag: delta => {
                    SNPCoverageDisplay.setHeight(SNPCoverageDisplay.height + delta);
                    return delta;
                }, className: classes.resizeHandle, style: {
                    top,
                } }))) : null,
        showPileup ? (react_1.default.createElement("div", { "data-testid": "Blockset-pileup", style: {
                position: 'absolute',
                top: showCoverage ? SNPCoverageDisplay.height : 0,
            } },
            react_1.default.createElement(PileupDisplay.RenderingComponent, { model: PileupDisplay }))) : null));
}
exports.default = (0, mobx_react_1.observer)(AlignmentsDisplay);
