// @jbrowse/plugin-variants/dist/VcfTabixAdapter/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mobx_state_tree_1 = require("mobx-state-tree");
const configuration_1 = require("@jbrowse/core/configuration");
/**
 * #config VcfTabixAdapter
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const VcfTabixAdapter = (0, configuration_1.ConfigurationSchema)('VcfTabixAdapter', {
    /**
     * #slot
     */
    vcfGzLocation: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/my.vcf.gz', locationType: 'UriLocation' },
    },
    index: (0, configuration_1.ConfigurationSchema)('VcfIndex', {
        /**
         * #slot index.indexType
         */
        indexType: {
            model: mobx_state_tree_1.types.enumeration('IndexType', ['TBI', 'CSI']),
            type: 'stringEnum',
            defaultValue: 'TBI',
        },
        /**
         * #slot index.location
         */
        location: {
            type: 'fileLocation',
            defaultValue: {
                uri: '/path/to/my.vcf.gz.tbi',
                locationType: 'UriLocation',
            },
        },
    }),
}, { explicitlyTyped: true });
exports.default = VcfTabixAdapter;
