// @jbrowse/plugin-svg/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SvgFeatureRendererReactComponent = exports.svgFeatureRendererConfigSchema = void 0;
const BoxRendererType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/renderers/BoxRendererType"));
const Plugin_1 = __importDefault(require("@jbrowse/core/Plugin"));
const SvgFeatureRenderer_1 = require("./SvgFeatureRenderer");
Object.defineProperty(exports, "svgFeatureRendererConfigSchema", { enumerable: true, get: function () { return SvgFeatureRenderer_1.configSchema; } });
Object.defineProperty(exports, "SvgFeatureRendererReactComponent", { enumerable: true, get: function () { return SvgFeatureRenderer_1.ReactComponent; } });
class SvgFeatureRenderer extends BoxRendererType_1.default {
    constructor() {
        super(...arguments);
        this.supportsSVG = true;
    }
}
class SVGPlugin extends Plugin_1.default {
    constructor() {
        super(...arguments);
        this.name = 'SVGPlugin';
    }
    install(pluginManager) {
        pluginManager.addRendererType(() => new SvgFeatureRenderer({
            name: 'SvgFeatureRenderer',
            ReactComponent: SvgFeatureRenderer_1.ReactComponent,
            configSchema: SvgFeatureRenderer_1.configSchema,
            pluginManager,
        }));
    }
}
exports.default = SVGPlugin;
