// @jbrowse/plugin-bed/dist/BedTabixAdapter/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const configuration_1 = require("@jbrowse/core/configuration");
const mobx_state_tree_1 = require("mobx-state-tree");
/**
 * #config BedTabixAdapter
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const BedTabixAdapter = (0, configuration_1.ConfigurationSchema)('BedTabixAdapter', {
    /**
     * #slot
     */
    bedGzLocation: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/my.bed.gz', locationType: 'UriLocation' },
    },
    index: (0, configuration_1.ConfigurationSchema)('TabixIndex', {
        /**
         * #slot index.indexType
         */
        indexType: {
            model: mobx_state_tree_1.types.enumeration('IndexType', ['TBI', 'CSI']),
            type: 'stringEnum',
            defaultValue: 'TBI',
        },
        /**
         * #slot index.location
         */
        location: {
            type: 'fileLocation',
            defaultValue: {
                uri: '/path/to/my.bed.gz.tbi',
                locationType: 'UriLocation',
            },
        },
    }),
    /**
     * #slot
     */
    columnNames: {
        type: 'stringArray',
        description: 'List of column names',
        defaultValue: [],
    },
    /**
     * #slot
     */
    scoreColumn: {
        type: 'string',
        description: 'The column to use as a "score" attribute',
        defaultValue: '',
    },
    /**
     * #slot
     */
    autoSql: {
        type: 'string',
        description: 'The autoSql definition for the data fields in the file',
        defaultValue: '',
    },
}, { explicitlyTyped: true });
exports.default = BedTabixAdapter;
