// @jbrowse/react-linear-genome-view/esm/JBrowseLinearGenomeView/AboutDialog.js
import React from 'react';
import { version } from '../version';
import { Dialog, DialogContent, DialogTitle, IconButton, Link, Typography, } from '@mui/material';
import CloseIcon from '@mui/icons-material/Close';
import { makeStyles } from 'tss-react/mui';
const useStyles = makeStyles()(theme => ({
    closeButton: {
        position: 'absolute',
        right: theme.spacing(1),
        top: theme.spacing(1),
        color: theme.palette.grey[500],
    },
    content: {
        minWidth: 800,
    },
}));
export default function AboutDialog({ open, onClose, }) {
    const { classes } = useStyles();
    return (React.createElement(Dialog, { open: open, onClose: onClose, maxWidth: "xl" },
        React.createElement(DialogTitle, null,
            "JBrowse v",
            version,
            React.createElement(IconButton, { className: classes.closeButton, onClick: () => onClose(), size: "large" },
                React.createElement(CloseIcon, null))),
        React.createElement(DialogContent, { className: classes.content },
            React.createElement(Typography, null, "JBrowse is a GMOD project \u00A9 2019-2021, The Evolutionary Software Foundation"),
            React.createElement(Typography, null, "Here are some resources and documentation. Please report the version number above when asking questions. Thanks!"),
            React.createElement("ul", null,
                React.createElement("li", null,
                    React.createElement(Link, { href: "https://github.com/GMOD/jbrowse-components/discussions", target: "_blank", rel: "noopener noreferrer" }, "Question & answer forum")),
                React.createElement("li", null,
                    React.createElement(Link, { href: "https://github.com/GMOD/jbrowse-components/issues/new/choose", target: "_blank", rel: "noopener noreferrer" }, "Report a bug")),
                React.createElement("li", null,
                    React.createElement(Link, { href: "https://jbrowse.org/jb2/docs/user_guide", target: "_blank", rel: "noopener noreferrer" }, "User guide")),
                React.createElement("li", null,
                    React.createElement(Link, { href: "https://jbrowse.org/jb2/docs/", target: "_blank", rel: "noopener noreferrer" }, "Documentation"))))));
}
