// @jbrowse/plugin-alignments/dist/SNPCoverageAdapter/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const AdapterType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/AdapterType"));
const configSchema_1 = __importDefault(require("./configSchema"));
const SNPCoverageAdapter_1 = require("./SNPCoverageAdapter");
function default_1(pluginManager) {
    pluginManager.addAdapterType(() => {
        return new AdapterType_1.default({
            name: 'SNPCoverageAdapter',
            adapterMetadata: {
                category: null,
                displayName: null,
                hiddenFromGUI: true,
                description: null,
            },
            getAdapterClass: () => Promise.resolve().then(() => __importStar(require('./SNPCoverageAdapter'))).then(r => r.default),
            configSchema: configSchema_1.default,
            adapterCapabilities: SNPCoverageAdapter_1.capabilities,
        });
    });
}
exports.default = default_1;
