// @jbrowse/plugin-wiggle/dist/LinearWiggleDisplay/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tooltip = exports.modelFactory = exports.ReactComponent = void 0;
const DisplayType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/DisplayType"));
const configSchema_1 = __importDefault(require("./models/configSchema"));
const model_1 = __importDefault(require("./models/model"));
exports.modelFactory = model_1.default;
const WiggleDisplayComponent_1 = __importDefault(require("./components/WiggleDisplayComponent"));
exports.ReactComponent = WiggleDisplayComponent_1.default;
const Tooltip_1 = __importDefault(require("./components/Tooltip"));
exports.Tooltip = Tooltip_1.default;
exports.default = (pluginManager) => {
    pluginManager.addDisplayType(() => {
        const configSchema = (0, configSchema_1.default)(pluginManager);
        const stateModel = (0, model_1.default)(pluginManager, configSchema);
        return new DisplayType_1.default({
            name: 'LinearWiggleDisplay',
            configSchema,
            stateModel,
            trackType: 'QuantitativeTrack',
            viewType: 'LinearGenomeView',
            ReactComponent: WiggleDisplayComponent_1.default,
        });
    });
};
