// @jbrowse/plugin-data-management/esm/HierarchicalTrackSelectorWidget/components/ManageConnectionsDialog.js
import React from 'react';
import { Button, Dialog, DialogTitle, DialogContent, DialogActions, IconButton, Tooltip, Typography, } from '@mui/material';
import { makeStyles } from 'tss-react/mui';
import CloseIcon from '@mui/icons-material/Close';
import { observer } from 'mobx-react';
import { readConfObject, } from '@jbrowse/core/configuration';
const useStyles = makeStyles()(theme => ({
    closeButton: {
        position: 'absolute',
        right: theme.spacing(1),
        top: theme.spacing(1),
        color: theme.palette.grey[500],
    },
    connectionContainer: {
        margin: theme.spacing(4),
        width: 500,
    },
}));
function ManageConnectionsDlg({ session, handleClose, breakConnection, }) {
    const { classes } = useStyles();
    const { adminMode, connections, sessionConnections } = session;
    return (React.createElement(Dialog, { open: true, onClose: handleClose, maxWidth: "lg" },
        React.createElement(DialogTitle, null,
            "Delete connections",
            React.createElement(IconButton, { className: classes.closeButton, onClick: () => handleClose() },
                React.createElement(CloseIcon, null))),
        React.createElement(DialogContent, null,
            React.createElement(Typography, null, "Click the X icon to delete the connection from your config completely"),
            React.createElement("div", { className: classes.connectionContainer },
                connections.map(conf => {
                    const name = readConfObject(conf, 'name');
                    return (React.createElement(Typography, { key: `conn-${name}` },
                        adminMode || (sessionConnections === null || sessionConnections === void 0 ? void 0 : sessionConnections.includes(conf)) ? (React.createElement(IconButton, { onClick: () => breakConnection(conf, true) },
                            React.createElement(CloseIcon, { color: "error" }))) : (React.createElement(Tooltip, { title: "Unable to delete connection in config file as non-admin user" },
                            React.createElement(IconButton, null,
                                React.createElement(CloseIcon, { color: "disabled" })))),
                        name));
                }),
                !connections.length ? (React.createElement(Typography, null, "No connections found")) : null)),
        React.createElement(DialogActions, null,
            React.createElement(Button, { onClick: () => handleClose(), variant: "contained", color: "primary" }, "Close"))));
}
export default observer(ManageConnectionsDlg);
