// @jbrowse/plugin-sequence/dist/DivSequenceRenderer/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const FeatureRendererType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/renderers/FeatureRendererType"));
const DivSequenceRendering_1 = __importDefault(require("./components/DivSequenceRendering"));
const configSchema_1 = __importDefault(require("./configSchema"));
/* adjust in both directions */
class DivSequenceRenderer extends FeatureRendererType_1.default {
    constructor() {
        super(...arguments);
        this.supportsSVG = true;
    }
    getExpandedRegion(region) {
        return {
            ...region,
            start: Math.max(region.start - 3, 0),
            end: region.end + 3,
        };
    }
}
exports.default = (pluginManager) => {
    pluginManager.addRendererType(() => new DivSequenceRenderer({
        name: 'DivSequenceRenderer',
        ReactComponent: DivSequenceRendering_1.default,
        configSchema: configSchema_1.default,
        pluginManager,
    }));
};
