// @jbrowse/plugin-variants/dist/extensionPoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tracks_1 = require("@jbrowse/core/util/tracks");
exports.default = (pluginManager) => {
    pluginManager.addToExtensionPoint('Core-guessAdapterForLocation', (adapterGuesser) => {
        return (file, index, adapterHint) => {
            const regexGuess = /\.vcf\.b?gz$/i;
            const adapterName = 'VcfTabixAdapter';
            const fileName = (0, tracks_1.getFileName)(file);
            const indexName = index && (0, tracks_1.getFileName)(index);
            const obj = {
                type: adapterName,
                vcfGzLocation: file,
                index: {
                    location: index || (0, tracks_1.makeIndex)(file, '.tbi'),
                    indexType: (0, tracks_1.makeIndexType)(indexName, 'CSI', 'TBI'),
                },
            };
            if (regexGuess.test(fileName) && !adapterHint) {
                return obj;
            }
            else if (adapterHint === adapterName) {
                return obj;
            }
            return adapterGuesser(file, index, adapterHint);
        };
    });
    pluginManager.addToExtensionPoint('Core-guessTrackTypeForLocation', (trackTypeGuesser) => {
        return (adapterName) => {
            if (adapterName === 'VcfTabixAdapter' || adapterName === 'VcfAdapter') {
                return 'VariantTrack';
            }
            return trackTypeGuesser(adapterName);
        };
    });
    pluginManager.addToExtensionPoint('Core-guessAdapterForLocation', (adapterGuesser) => {
        return (file, index, adapterHint) => {
            const regexGuess = /\.vcf$/i;
            const adapterName = 'VcfAdapter';
            const fileName = (0, tracks_1.getFileName)(file);
            if (regexGuess.test(fileName) || adapterHint === adapterName) {
                return {
                    type: adapterName,
                    vcfLocation: file,
                };
            }
            return adapterGuesser(file, index, adapterHint);
        };
    });
};
