// @jbrowse/plugin-alignments/esm/LinearPileupDisplay/components/ColorByTag.js
import React, { useState } from 'react';
import { observer } from 'mobx-react';
import { Button, Dialog, DialogContent, DialogActions, DialogTitle, IconButton, TextField, Typography, } from '@mui/material';
import { makeStyles } from 'tss-react/mui';
import CloseIcon from '@mui/icons-material/Close';
const useStyles = makeStyles()(theme => ({
    root: {
        width: 300,
    },
    closeButton: {
        position: 'absolute',
        right: theme.spacing(1),
        top: theme.spacing(1),
        color: theme.palette.grey[500],
    },
}));
function ColorByTagDlg(props) {
    const { classes } = useStyles();
    const { model, handleClose } = props;
    const [tag, setTag] = useState('');
    const validTag = tag.match(/^[A-Za-z][A-Za-z0-9]$/);
    return (React.createElement(Dialog, { open: true, onClose: handleClose },
        React.createElement(DialogTitle, null,
            "Color by tag",
            React.createElement(IconButton, { "aria-label": "close", className: classes.closeButton, onClick: handleClose },
                React.createElement(CloseIcon, null))),
        React.createElement(DialogContent, { style: { overflowX: 'hidden' } },
            React.createElement(Typography, null, "Enter tag to color by: "),
            React.createElement(Typography, { color: "textSecondary" }, "Examples: XS or TS for RNA-seq inferred read strand, ts (lower-case) for minimap2 read strand, HP for haplotype, RG for read group, etc."),
            React.createElement(TextField, { value: tag, onChange: event => setTag(event.target.value), placeholder: "Enter tag name", inputProps: {
                    maxLength: 2,
                    'data-testid': 'color-tag-name-input',
                }, error: tag.length === 2 && !validTag, helperText: tag.length === 2 && !validTag ? 'Not a valid tag' : '', autoComplete: "off", "data-testid": "color-tag-name" }),
            React.createElement(DialogActions, null,
                React.createElement(Button, { variant: "contained", color: "primary", onClick: () => {
                        model.setColorScheme({
                            type: 'tag',
                            tag,
                        });
                        handleClose();
                    }, disabled: !validTag }, "Submit"),
                React.createElement(Button, { variant: "contained", color: "secondary", onClick: handleClose }, "Cancel")))));
}
export default observer(ColorByTagDlg);
