// @jbrowse/plugin-trix/dist/TrixTextSearchAdapter/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const configuration_1 = require("@jbrowse/core/configuration");
/**
 * #config TrixTextSearchAdapter
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const TrixTextSearchAdapter = (0, configuration_1.ConfigurationSchema)('TrixTextSearchAdapter', {
    /**
     * #slot
     */
    ixFilePath: {
        type: 'fileLocation',
        defaultValue: { uri: 'out.ix', locationType: 'UriLocation' },
        description: 'the location of the trix ix file',
    },
    /**
     * #slot
     */
    ixxFilePath: {
        type: 'fileLocation',
        defaultValue: { uri: 'out.ixx', locationType: 'UriLocation' },
        description: 'the location of the trix ixx file',
    },
    /**
     * #slot
     */
    metaFilePath: {
        type: 'fileLocation',
        defaultValue: { uri: 'meta.json', locationType: 'UriLocation' },
        description: 'the location of the metadata json file for the trix index',
    },
    /**
     * #slot
     */
    tracks: {
        type: 'stringArray',
        defaultValue: [],
        description: 'List of tracks covered by text search adapter',
    },
    /**
     * #slot
     */
    assemblyNames: {
        type: 'stringArray',
        defaultValue: [],
        description: 'List of assemblies covered by text search adapter',
    },
}, {
    explicitlyTyped: true,
    /**
     * #identifier
     */
    explicitIdentifier: 'textSearchAdapterId',
});
exports.default = TrixTextSearchAdapter;
