// @jbrowse/plugin-legacy-jbrowse/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const AdapterType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/AdapterType"));
const TextSearchAdapterType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/TextSearchAdapterType"));
const ConnectionType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/ConnectionType"));
const Plugin_1 = __importDefault(require("@jbrowse/core/Plugin"));
const NCListAdapter_1 = require("./NCListAdapter");
const JBrowse1TextSeachAdapter_1 = require("./JBrowse1TextSeachAdapter");
const tracks_1 = require("@jbrowse/core/util/tracks");
const JBrowse1Connection_1 = require("./JBrowse1Connection");
class LegacyJBrowsePlugin extends Plugin_1.default {
    constructor() {
        super(...arguments);
        this.name = 'LegacyJBrowsePlugin';
    }
    install(pluginManager) {
        pluginManager.addAdapterType(() => new AdapterType_1.default({
            name: 'NCListAdapter',
            configSchema: NCListAdapter_1.configSchema,
            getAdapterClass: () => Promise.resolve().then(() => __importStar(require('./NCListAdapter/NCListAdapter'))).then(r => r.default),
        }));
        pluginManager.addToExtensionPoint('Core-guessAdapterForLocation', (adapterGuesser) => {
            return (file, index, adapterHint) => {
                const regexGuess = /trackData.jsonz?$/i;
                const adapterName = 'NCListAdapter';
                const fileName = (0, tracks_1.getFileName)(file);
                if (regexGuess.test(fileName) || adapterHint === adapterName) {
                    return {
                        type: adapterName,
                        rootUrlTemplate: file,
                    };
                }
                return adapterGuesser(file, index, adapterHint);
            };
        });
        pluginManager.addTextSearchAdapterType(() => new TextSearchAdapterType_1.default({
            name: 'JBrowse1TextSearchAdapter',
            configSchema: JBrowse1TextSeachAdapter_1.configSchema,
            AdapterClass: JBrowse1TextSeachAdapter_1.AdapterClass,
            description: 'A JBrowse 1 text search adapter',
        }));
        pluginManager.addConnectionType(() => new ConnectionType_1.default({
            name: 'JBrowse1Connection',
            configSchema: JBrowse1Connection_1.configSchema,
            stateModel: (0, JBrowse1Connection_1.modelFactory)(pluginManager),
            displayName: 'JBrowse 1 Data',
            description: 'A JBrowse 1 data directory',
            url: '//jbrowse.org/',
        }));
    }
}
exports.default = LegacyJBrowsePlugin;
