// @jbrowse/plugin-data-management/dist/AddTrackWidget/components/TrackSourceSelect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const ui_1 = require("@jbrowse/core/ui");
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const mobx_state_tree_1 = require("mobx-state-tree");
const mobx_react_1 = require("mobx-react");
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    paper: {
        display: 'flex',
        flexDirection: 'column',
        padding: theme.spacing(1),
    },
    spacer: {
        height: theme.spacing(8),
    },
}));
function TrackSourceSelect({ model }) {
    const { classes } = useStyles();
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rootModel = (0, mobx_state_tree_1.getRoot)(model);
    return (react_1.default.createElement(material_1.Paper, { className: classes.paper },
        react_1.default.createElement(ui_1.FileSelector, { name: "Main file", description: "", location: model.trackData, setLocation: model.setTrackData, setName: model.setTrackName, rootModel: rootModel }),
        react_1.default.createElement("div", { className: classes.spacer }),
        react_1.default.createElement(ui_1.FileSelector, { name: "Index file", description: "(Optional) The URL of the index file is automatically inferred from the URL of the main file if it is not supplied.", location: model.indexTrackData, setLocation: model.setIndexTrackData, setName: model.setTrackName, rootModel: rootModel })));
}
exports.default = (0, mobx_react_1.observer)(TrackSourceSelect);
