// @jbrowse/plugin-linear-genome-view/dist/BaseLinearDisplay/components/Tooltip.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const configuration_1 = require("@jbrowse/core/configuration");
const ui_1 = require("@jbrowse/core/ui");
const mobx_react_1 = require("mobx-react");
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const react_popper_1 = require("react-popper");
function round(value) {
    return Math.round(value * 1e5) / 1e5;
}
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    // these styles come from
    // https://github.com/mui-org/material-ui/blob/master/packages/material-ui/src/Tooltip/Tooltip.js
    tooltip: {
        pointerEvents: 'none',
        backgroundColor: (0, material_1.alpha)(theme.palette.grey[700], 0.9),
        borderRadius: theme.shape.borderRadius,
        color: theme.palette.common.white,
        fontFamily: theme.typography.fontFamily,
        padding: '4px 8px',
        fontSize: theme.typography.pxToRem(12),
        lineHeight: `${round(14 / 10)}em`,
        maxWidth: 300,
        wordWrap: 'break-word',
    },
}));
const TooltipContents = react_1.default.forwardRef(({ message }, ref) => {
    return (react_1.default.createElement("div", { ref: ref }, react_1.default.isValidElement(message) ? (message) : message ? (react_1.default.createElement(ui_1.SanitizedHTML, { html: String(message) })) : null));
});
const Tooltip = (0, mobx_react_1.observer)(({ model, clientMouseCoord, }) => {
    const { classes } = useStyles();
    const { featureUnderMouse } = model;
    const [width, setWidth] = (0, react_1.useState)(0);
    const [popperElt, setPopperElt] = (0, react_1.useState)(null);
    // must be memoized a la https://github.com/popperjs/react-popper/issues/391
    const virtElement = (0, react_1.useMemo)(() => ({
        getBoundingClientRect: () => {
            const x = clientMouseCoord[0] + width / 2 + 20;
            const y = clientMouseCoord[1];
            return {
                top: y,
                left: x,
                bottom: y,
                right: x,
                width: 0,
                height: 0,
                x,
                y,
                toJSON() { },
            };
        },
    }), [clientMouseCoord, width]);
    const { styles, attributes } = (0, react_popper_1.usePopper)(virtElement, popperElt);
    const contents = featureUnderMouse
        ? (0, configuration_1.getConf)(model, 'mouseover', { feature: featureUnderMouse })
        : undefined;
    return featureUnderMouse && contents ? (react_1.default.createElement(material_1.Portal, null,
        react_1.default.createElement("div", { ref: setPopperElt, className: classes.tooltip, 
            // zIndex needed to go over widget drawer
            style: { ...styles.popper, zIndex: 100000 }, ...attributes.popper },
            react_1.default.createElement(TooltipContents, { ref: elt => setWidth((elt === null || elt === void 0 ? void 0 : elt.getBoundingClientRect().width) || 0), message: contents })))) : null;
});
exports.default = Tooltip;
