// @jbrowse/react-linear-genome-view/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPlugins = exports.createViewState = exports.createModel = exports.JBrowseLinearGenomeView = exports.ThemeProvider = exports.createJBrowseTheme = void 0;
var deprecations_1 = require("./deprecations");
Object.defineProperty(exports, "createJBrowseTheme", { enumerable: true, get: function () { return deprecations_1.createJBrowseTheme; } });
Object.defineProperty(exports, "ThemeProvider", { enumerable: true, get: function () { return deprecations_1.ThemeProvider; } });
var JBrowseLinearGenomeView_1 = require("./JBrowseLinearGenomeView");
Object.defineProperty(exports, "JBrowseLinearGenomeView", { enumerable: true, get: function () { return __importDefault(JBrowseLinearGenomeView_1).default; } });
var createModel_1 = require("./createModel");
Object.defineProperty(exports, "createModel", { enumerable: true, get: function () { return __importDefault(createModel_1).default; } });
var createViewState_1 = require("./createViewState");
Object.defineProperty(exports, "createViewState", { enumerable: true, get: function () { return __importDefault(createViewState_1).default; } });
var loadPlugins_1 = require("./loadPlugins");
Object.defineProperty(exports, "loadPlugins", { enumerable: true, get: function () { return __importDefault(loadPlugins_1).default; } });
