// @jbrowse/plugin-circular-view/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseChordDisplayComponent = exports.baseChordDisplayConfig = exports.BaseChordDisplayModel = void 0;
const react_1 = require("react");
const mobx_1 = require("mobx");
const util_1 = require("@jbrowse/core/util");
const Plugin_1 = __importDefault(require("@jbrowse/core/Plugin"));
const ViewType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/ViewType"));
const DataUsage_1 = __importDefault(require("@mui/icons-material/DataUsage"));
const CircularView_1 = __importDefault(require("./CircularView/models/CircularView"));
class CircularViewPlugin extends Plugin_1.default {
    constructor() {
        super(...arguments);
        this.name = 'CircularViewPlugin';
    }
    install(pluginManager) {
        pluginManager.addViewType(() => new ViewType_1.default({
            ReactComponent: (0, react_1.lazy)(() => Promise.resolve().then(() => __importStar(require('./CircularView/components/CircularView')))),
            stateModel: (0, CircularView_1.default)(pluginManager),
            name: 'CircularView',
        }));
        pluginManager.addToExtensionPoint('LaunchView-CircularView', 
        // @ts-ignore
        async ({ session, assembly, loc, tracks = [], }) => {
            const { assemblyManager } = session;
            const view = session.addView('CircularView', {});
            await (0, mobx_1.when)(() => view.initialized);
            if (!assembly) {
                throw new Error('No assembly provided when launching circular genome view');
            }
            const asm = await assemblyManager.waitForAssembly(assembly);
            if (!asm) {
                throw new Error(`Assembly "${assembly}" not found when launching circular genome view`);
            }
            view.setDisplayedRegions(asm.regions || []);
            tracks.forEach(track => view.showTrack(track));
        });
    }
    configure(pluginManager) {
        if ((0, util_1.isAbstractMenuManager)(pluginManager.rootModel)) {
            pluginManager.rootModel.appendToSubMenu(['Add'], {
                label: 'Circular view',
                icon: DataUsage_1.default,
                onClick: (session) => {
                    session.addView('CircularView', {});
                },
            });
        }
    }
}
exports.default = CircularViewPlugin;
var BaseChordDisplay_1 = require("./BaseChordDisplay");
Object.defineProperty(exports, "BaseChordDisplayModel", { enumerable: true, get: function () { return BaseChordDisplay_1.BaseChordDisplayModel; } });
Object.defineProperty(exports, "baseChordDisplayConfig", { enumerable: true, get: function () { return BaseChordDisplay_1.baseChordDisplayConfig; } });
Object.defineProperty(exports, "BaseChordDisplayComponent", { enumerable: true, get: function () { return BaseChordDisplay_1.BaseChordDisplayComponent; } });
