// @jbrowse/plugin-wiggle/dist/LinearWiggleDisplay/models/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const configuration_1 = require("@jbrowse/core/configuration");
const plugin_linear_genome_view_1 = require("@jbrowse/plugin-linear-genome-view");
const mobx_state_tree_1 = require("mobx-state-tree");
/**
 * #config LinearWiggleDisplay
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
function WiggleConfigFactory(pluginManager) {
    const XYPlotRendererConfigSchema = pluginManager.getRendererType('XYPlotRenderer').configSchema;
    const DensityRendererConfigSchema = pluginManager.getRendererType('DensityRenderer').configSchema;
    const LinePlotRendererConfigSchema = pluginManager.getRendererType('LinePlotRenderer').configSchema;
    return (0, configuration_1.ConfigurationSchema)('LinearWiggleDisplay', {
        /**
         * #slot
         */
        autoscale: {
            type: 'stringEnum',
            defaultValue: 'local',
            model: mobx_state_tree_1.types.enumeration('Autoscale type', [
                'global',
                'local',
                'globalsd',
                'localsd',
                'zscore',
            ]),
            description: 'global/local using their min/max values or w/ standard deviations (globalsd/localsd)',
        },
        /**
         * #slot
         */
        minimalTicks: {
            type: 'boolean',
            defaultValue: false,
            description: 'use the minimal amount of ticks',
        },
        /**
         * #slot
         */
        minScore: {
            type: 'number',
            defaultValue: Number.MIN_VALUE,
            description: 'minimum value for the y-scale',
        },
        /**
         * #slot
         */
        maxScore: {
            type: 'number',
            description: 'maximum value for the y-scale',
            defaultValue: Number.MAX_VALUE,
        },
        /**
         * #slot
         */
        numStdDev: {
            type: 'number',
            description: 'number of standard deviations to use for autoscale types globalsd or localsd',
            defaultValue: 3,
        },
        /**
         * #slot
         */
        scaleType: {
            type: 'stringEnum',
            model: mobx_state_tree_1.types.enumeration('Scale type', ['linear', 'log']),
            description: 'The type of scale to use',
            defaultValue: 'linear',
        },
        /**
         * #slot
         */
        inverted: {
            type: 'boolean',
            description: 'draw upside down',
            defaultValue: false,
        },
        /**
         * #slot
         */
        defaultRendering: {
            type: 'stringEnum',
            model: mobx_state_tree_1.types.enumeration('Rendering', ['density', 'xyplot', 'line']),
            defaultValue: 'xyplot',
        },
        /**
         * #slot
         */
        renderers: (0, configuration_1.ConfigurationSchema)('RenderersConfiguration', {
            DensityRenderer: DensityRendererConfigSchema,
            XYPlotRenderer: XYPlotRendererConfigSchema,
            LinePlotRenderer: LinePlotRendererConfigSchema,
        }),
    }, {
        /**
         * #baseConfiguration
         */
        baseConfiguration: plugin_linear_genome_view_1.baseLinearDisplayConfigSchema,
        explicitlyTyped: true,
    });
}
exports.default = WiggleConfigFactory;
