// @jbrowse/plugin-circular-view/dist/BaseChordDisplay/components/Loading.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const mui_1 = require("tss-react/mui");
const mobx_react_1 = require("mobx-react");
const useStyles = (0, mui_1.makeStyles)()(theme => {
    const offset = 2;
    const duration = 1.4;
    const { primary, secondary, tertiary, quaternary } = theme.palette;
    return {
        path: {
            strokeDasharray: 187,
            strokeDashoffset: 50,
            animation: `$dash ${duration}s ease-in-out infinite, $colors ${duration * 4}s ease-in-out infinite`,
        },
        '@keyframes colors': {
            '0%': {
                stroke: primary.light,
            },
            '25%': {
                stroke: secondary.light,
            },
            '50%': {
                stroke: tertiary.light,
            },
            '75%': {
                stroke: quaternary.light,
            },
            '100%': {
                stroke: primary.light,
            },
        },
        '@keyframes dash': {
            '0%': {
                strokeDashoffset: offset,
            },
            '50%': {
                strokeDashoffset: offset / 4,
                transform: 'rotate(135deg)',
            },
            '100%': {
                strokeDashoffset: offset,
                transform: 'rotate(720deg)',
            },
        },
    };
});
// 'repeating-linear-gradient(45deg, transparent, transparent 5px, rgba(255,255,255,.5) 5px, rgba(255,255,255,.5) 10px)',
const Loading = (0, mobx_react_1.observer)(({ model: { renderProps: { radius }, }, }) => {
    const { classes } = useStyles();
    // only show the loading message after 400ms to prevent excessive flickering
    const [shown, setShown] = (0, react_1.useState)(false);
    (0, react_1.useEffect)(() => {
        const timeout = setTimeout(() => setShown(true), 400);
        return () => clearTimeout(timeout);
    });
    return !shown ? null : (react_1.default.createElement("g", null,
        react_1.default.createElement("defs", null,
            react_1.default.createElement("pattern", { id: "diagonalHatch", width: "10", height: "10", patternTransform: "rotate(45 0 0)", patternUnits: "userSpaceOnUse" },
                react_1.default.createElement("line", { x1: "0", y1: "0", x2: "0", y2: "10", style: { stroke: 'rgba(255,255,255,0.5)', strokeWidth: 10 } }))),
        react_1.default.createElement("circle", { cx: "0", cy: "0", r: radius, fill: "#f1f1f1" }),
        react_1.default.createElement("circle", { cx: "0", cy: "0", r: radius, fill: "url(#diagonalHatch)" }),
        react_1.default.createElement("text", { x: "0", y: "0", transform: "rotate(90 0 0)", dominantBaseline: "middle", textAnchor: "middle" }, "Loading\u2026"),
        react_1.default.createElement("circle", { className: classes.path, fill: "none", strokeWidth: "4", strokeLinecap: "round", cx: "0", cy: "0", r: "60" })));
});
exports.default = Loading;
