// @jbrowse/core/ui/Drawer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const mobx_react_1 = require("mobx-react");
const ResizeHandle_1 = __importDefault(require("./ResizeHandle"));
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    paper: {
        overflowY: 'auto',
        height: '100%',
        zIndex: theme.zIndex.drawer,
        outline: 'none',
        background: theme.palette.background.default,
    },
    resizeHandle: {
        width: 4,
        position: 'fixed',
        top: 0,
        zIndex: theme.zIndex.drawer + 1,
    },
}));
function Drawer({ children, session, }) {
    const { drawerPosition, drawerWidth } = session;
    const { classes } = useStyles();
    return (react_1.default.createElement(material_1.Paper, { className: classes.paper, elevation: 16, square: true },
        drawerPosition === 'right' ? (react_1.default.createElement(ResizeHandle_1.default, { onDrag: session.resizeDrawer, className: classes.resizeHandle, vertical: true })) : null,
        children,
        drawerPosition === 'left' ? (react_1.default.createElement(ResizeHandle_1.default, { onDrag: session.resizeDrawer, className: classes.resizeHandle, style: { left: drawerWidth }, vertical: true })) : null));
}
exports.default = (0, mobx_react_1.observer)(Drawer);
