// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/HelpDialog.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const Close_1 = __importDefault(require("@mui/icons-material/Close"));
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    closeButton: {
        position: 'absolute',
        right: theme.spacing(1),
        top: theme.spacing(1),
        color: theme.palette.grey[500],
    },
}));
function HelpDialog({ handleClose, }) {
    const { classes } = useStyles();
    return (react_1.default.createElement(material_1.Dialog, { open: true, maxWidth: "xl", onClose: handleClose },
        react_1.default.createElement(material_1.DialogTitle, null,
            "Using the search box",
            handleClose ? (react_1.default.createElement(material_1.IconButton, { className: classes.closeButton, onClick: () => handleClose() },
                react_1.default.createElement(Close_1.default, null))) : null),
        react_1.default.createElement(material_1.Divider, null),
        react_1.default.createElement(material_1.DialogContent, null,
            react_1.default.createElement("h3", null, "Searching"),
            react_1.default.createElement("ul", null,
                react_1.default.createElement("li", null, "Jump to a feature or reference sequence by typing its name in the location box and pressing Enter."),
                react_1.default.createElement("li", null,
                    "Jump to a specific region by typing the region into the location box as: ",
                    react_1.default.createElement("code", null, "ref:start..end"),
                    " or ",
                    react_1.default.createElement("code", null, "ref:start-end"),
                    ". Commas are allowed in the start and end coordinates. A space-separated list of locstrings can be used to open up multiple chromosomes at a time")),
            react_1.default.createElement("h3", null, "Example Searches"),
            react_1.default.createElement("ul", null,
                react_1.default.createElement("li", null,
                    react_1.default.createElement("code", null, "BRCA"),
                    " - searches for the feature named BRCA"),
                react_1.default.createElement("li", null,
                    react_1.default.createElement("code", null, "chr4"),
                    " - jumps to chromosome 4"),
                react_1.default.createElement("li", null,
                    react_1.default.createElement("code", null, "chr4:79,500,000..80,000,000"),
                    " - jumps the region on chromosome 4 between 79.5Mb and 80Mb."),
                react_1.default.createElement("li", null,
                    react_1.default.createElement("code", null, "chr1:1-100 chr2:1-100"),
                    " - create a split view of chr1:1-100 and chr2:1-100"),
                react_1.default.createElement("li", null,
                    react_1.default.createElement("code", null, "chr1 chr2 chr3"),
                    " - open up multiple chromosomes at once"),
                react_1.default.createElement("li", null,
                    react_1.default.createElement("code", null, "chr1:1-100[rev] chr2:1-100"),
                    " - open up the first region in the horizontally flipped orientation"),
                react_1.default.createElement("li", null,
                    react_1.default.createElement("code", null, "chr1 100 200"),
                    " - use whitespace separated refname, start, end"))),
        react_1.default.createElement(material_1.Divider, null),
        react_1.default.createElement(material_1.DialogActions, null,
            react_1.default.createElement(material_1.Button, { onClick: () => handleClose(), color: "primary" }, "Close"))));
}
exports.default = HelpDialog;
