// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/ImportForm.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const mui_1 = require("tss-react/mui");
const mobx_react_1 = require("mobx-react");
const util_1 = require("@jbrowse/core/util");
const material_1 = require("@mui/material");
const ui_1 = require("@jbrowse/core/ui");
// icons
const Close_1 = __importDefault(require("@mui/icons-material/Close"));
// locals
const RefNameAutocomplete_1 = __importDefault(require("./RefNameAutocomplete"));
const util_2 = require("./util");
const SearchResultsDialog = (0, react_1.lazy)(() => Promise.resolve().then(() => __importStar(require('./SearchResultsDialog'))));
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    importFormContainer: {
        padding: theme.spacing(2),
    },
    button: {
        margin: theme.spacing(2),
    },
    container: {
        padding: theme.spacing(4),
    },
}));
const ImportForm = (0, mobx_react_1.observer)(({ model }) => {
    var _a;
    const { classes } = useStyles();
    const session = (0, util_1.getSession)(model);
    const { assemblyNames, assemblyManager, textSearchManager } = session;
    const { rankSearchResults, isSearchDialogDisplayed, error } = model;
    const [selectedAsm, setSelectedAsm] = (0, react_1.useState)(assemblyNames[0]);
    const [option, setOption] = (0, react_1.useState)();
    const searchScope = model.searchScope(selectedAsm);
    const assembly = assemblyManager.get(selectedAsm);
    const assemblyError = assemblyNames.length
        ? assembly === null || assembly === void 0 ? void 0 : assembly.error
        : 'No configured assemblies';
    const regions = (assembly === null || assembly === void 0 ? void 0 : assembly.regions) || [];
    const displayError = assemblyError || error;
    const [value, setValue] = (0, react_1.useState)('');
    const r0 = (_a = regions[0]) === null || _a === void 0 ? void 0 : _a.refName;
    // useEffect resets to an "initial state" of displaying first region from assembly
    // after assembly change. needs to react to selectedAsm as well as r0 because changing
    // assembly will run setValue('') and then r0 may not change if assembly names are the
    // same across assemblies, but it still needs to be reset
    (0, react_1.useEffect)(() => {
        setValue(r0);
    }, [r0, selectedAsm]);
    function navToOption(option) {
        const location = option.getLocation();
        const trackId = option.getTrackId();
        if (location) {
            model.navToLocString(location, selectedAsm);
            if (trackId) {
                model.showTrack(trackId);
            }
        }
    }
    // gets a string as input, or use stored option results from previous query,
    // then re-query and
    // 1) if it has multiple results: pop a dialog
    // 2) if it's a single result navigate to it
    // 3) else assume it's a locstring and navigate to it
    async function handleSelectedRegion(input) {
        try {
            if ((option === null || option === void 0 ? void 0 : option.getDisplayString()) === input && option.hasLocation()) {
                navToOption(option);
            }
            else {
                const [ref, rest] = (0, util_2.splitLast)(input, ':');
                const allRefs = (assembly === null || assembly === void 0 ? void 0 : assembly.allRefNamesWithLowerCase) || [];
                if (allRefs.includes(input) ||
                    (allRefs.includes(ref) && !Number.isNaN(parseInt(rest, 10)))) {
                    model.navToLocString(input, selectedAsm);
                }
                else {
                    const results = await (0, util_2.fetchResults)({
                        queryString: input,
                        searchType: 'exact',
                        searchScope,
                        rankSearchResults,
                        textSearchManager,
                        assembly,
                    });
                    if (results.length > 1) {
                        model.setSearchResults(results, input.toLowerCase());
                    }
                    else if (results.length === 1) {
                        navToOption(results[0]);
                    }
                    else {
                        model.navToLocString(input, selectedAsm);
                    }
                }
            }
        }
        catch (e) {
            console.error(e);
            session.notify(`${e}`, 'warning');
        }
    }
    // implementation notes:
    // having this wrapped in a form allows intuitive use of enter key to submit
    return (react_1.default.createElement("div", { className: classes.container },
        displayError ? react_1.default.createElement(ui_1.ErrorMessage, { error: displayError }) : null,
        react_1.default.createElement(material_1.Container, { className: classes.importFormContainer },
            react_1.default.createElement("form", { onSubmit: event => {
                    event.preventDefault();
                    model.setError(undefined);
                    if (value) {
                        handleSelectedRegion(value);
                    }
                } },
                react_1.default.createElement(material_1.Grid, { container: true, spacing: 1, justifyContent: "center", alignItems: "center" },
                    react_1.default.createElement(material_1.Grid, { item: true },
                        react_1.default.createElement(material_1.FormControl, null,
                            react_1.default.createElement(ui_1.AssemblySelector, { onChange: val => {
                                    setSelectedAsm(val);
                                    setValue('');
                                }, session: session, selected: selectedAsm }))),
                    react_1.default.createElement(material_1.Grid, { item: true }, selectedAsm ? (assemblyError ? (react_1.default.createElement(Close_1.default, { style: { color: 'red' } })) : value ? (react_1.default.createElement(material_1.FormControl, null,
                        react_1.default.createElement(RefNameAutocomplete_1.default, { fetchResults: queryString => (0, util_2.fetchResults)({
                                queryString,
                                assembly,
                                textSearchManager,
                                rankSearchResults,
                                searchScope,
                            }), model: model, assemblyName: assemblyError ? undefined : selectedAsm, value: value, 
                            // note: minWidth 270 accomodates full width of helperText
                            minWidth: 270, onChange: str => setValue(str), onSelect: val => setOption(val), TextFieldProps: {
                                variant: 'outlined',
                                helperText: 'Enter sequence name, feature name, or location',
                                style: { minWidth: '175px' },
                            } }))) : (react_1.default.createElement(material_1.CircularProgress, { role: "progressbar", size: 20, disableShrink: true }))) : null),
                    react_1.default.createElement(material_1.Grid, { item: true },
                        react_1.default.createElement(material_1.FormControl, null,
                            react_1.default.createElement(material_1.Button, { type: "submit", disabled: !value, className: classes.button, variant: "contained", color: "primary" }, "Open")),
                        react_1.default.createElement(material_1.FormControl, null,
                            react_1.default.createElement(material_1.Button, { disabled: !value, className: classes.button, onClick: () => {
                                    model.setError(undefined);
                                    model.showAllRegionsInAssembly(selectedAsm);
                                }, variant: "contained", color: "secondary" }, "Show all regions in assembly")))))),
        isSearchDialogDisplayed ? (react_1.default.createElement(SearchResultsDialog, { model: model, optAssemblyName: selectedAsm, handleClose: () => model.setSearchResults(undefined, undefined) })) : null));
});
exports.default = ImportForm;
