// @jbrowse/plugin-linear-genome-view/dist/LinearGenomeView/components/OverviewRubberBand.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importStar(require("react"));
const material_1 = require("@mui/material");
const mui_1 = require("tss-react/mui");
const util_1 = require("@jbrowse/core/util");
const mobx_react_1 = require("mobx-react");
const __1 = require("..");
const useStyles = (0, mui_1.makeStyles)()(theme => {
    const { tertiary, primary } = theme.palette;
    const background = tertiary
        ? (0, material_1.alpha)(tertiary.main, 0.7)
        : (0, material_1.alpha)(primary.main, 0.7);
    return {
        rubberBand: {
            height: '100%',
            background,
            position: 'absolute',
            zIndex: 10,
            textAlign: 'center',
            overflow: 'hidden',
        },
        rubberBandControl: {
            cursor: 'crosshair',
            width: '100%',
            minHeight: 8,
        },
        rubberBandText: {
            color: tertiary ? tertiary.contrastText : primary.contrastText,
        },
        popover: {
            mouseEvents: 'none',
            cursor: 'crosshair',
        },
        paper: {
            paddingLeft: theme.spacing(1),
            paddingRight: theme.spacing(1),
        },
        guide: {
            pointerEvents: 'none',
            height: '100%',
            width: 1,
            position: 'absolute',
            zIndex: 10,
        },
    };
});
const HoverTooltip = (0, mobx_react_1.observer)(({ model, open, guideX, overview, }) => {
    var _a;
    const { classes } = useStyles();
    const { cytobandOffset } = model;
    const { assemblyManager } = (0, util_1.getSession)(model);
    const px = overview.pxToBp(guideX - cytobandOffset);
    const assembly = assemblyManager.get(px.assemblyName);
    const cytoband = (_a = assembly === null || assembly === void 0 ? void 0 : assembly.cytobands) === null || _a === void 0 ? void 0 : _a.find(f => px.coord > f.get('start') &&
        px.coord < f.get('end') &&
        px.refName === assembly.getCanonicalRefName(f.get('refName')));
    return (react_1.default.createElement(material_1.Tooltip, { open: open, placement: "top", title: [(0, util_1.stringify)(px), cytoband === null || cytoband === void 0 ? void 0 : cytoband.get('name')].join(' '), arrow: true },
        react_1.default.createElement("div", { className: classes.guide, style: {
                left: guideX,
            } })));
});
function OverviewRubberBand({ model, overview, ControlComponent = react_1.default.createElement("div", null), }) {
    const { cytobandOffset } = model;
    const [startX, setStartX] = (0, react_1.useState)();
    const [currentX, setCurrentX] = (0, react_1.useState)();
    const [guideX, setGuideX] = (0, react_1.useState)();
    const controlsRef = (0, react_1.useRef)(null);
    const rubberBandRef = (0, react_1.useRef)(null);
    const { classes } = useStyles();
    const mouseDragging = startX !== undefined;
    (0, react_1.useEffect)(() => {
        function globalMouseMove(event) {
            const ref = controlsRef.current;
            if (ref && mouseDragging) {
                const relativeX = event.clientX - ref.getBoundingClientRect().left;
                setCurrentX(relativeX);
            }
        }
        function globalMouseUp() {
            // click and drag
            if (startX !== undefined && currentX !== undefined) {
                if (Math.abs(currentX - startX) > 3) {
                    const left = Math.min(startX, currentX);
                    const right = Math.max(startX, currentX);
                    model.moveTo(overview.pxToBp(left - cytobandOffset), overview.pxToBp(right - cytobandOffset));
                }
            }
            // just a click
            if (startX !== undefined && currentX === undefined) {
                const click = overview.pxToBp(startX - cytobandOffset);
                if (!click.refName) {
                    (0, util_1.getSession)(model).notify('unknown position clicked');
                    console.error('unknown position clicked', click);
                }
                else {
                    model.centerAt(Math.round(click.coord), click.refName, click.index);
                }
            }
            setStartX(undefined);
            setCurrentX(undefined);
            if (startX !== undefined) {
                setGuideX(undefined);
            }
        }
        function globalKeyDown(event) {
            if (event.key === 'Escape') {
                setStartX(undefined);
                setCurrentX(undefined);
            }
        }
        if (mouseDragging) {
            window.addEventListener('mousemove', globalMouseMove, true);
            window.addEventListener('mouseup', globalMouseUp, true);
            window.addEventListener('keydown', globalKeyDown, true);
            return () => {
                window.removeEventListener('mousemove', globalMouseMove, true);
                window.removeEventListener('mouseup', globalMouseUp, true);
                window.removeEventListener('keydown', globalKeyDown, true);
            };
        }
        return () => { };
    }, [mouseDragging, currentX, startX, model, overview, cytobandOffset]);
    function mouseDown(event) {
        event.preventDefault();
        event.stopPropagation();
        if (controlsRef.current) {
            setStartX(event.clientX - controlsRef.current.getBoundingClientRect().left);
        }
    }
    function mouseMove(event) {
        if (controlsRef.current) {
            setGuideX(event.clientX - controlsRef.current.getBoundingClientRect().left);
        }
    }
    function mouseOut() {
        setGuideX(undefined);
    }
    if (startX === undefined) {
        return (react_1.default.createElement("div", { style: { position: 'relative' } },
            guideX !== undefined ? (react_1.default.createElement(HoverTooltip, { model: model, open: !mouseDragging, overview: overview, guideX: guideX })) : null,
            react_1.default.createElement("div", { className: classes.rubberBandControl, role: "presentation", ref: controlsRef, onMouseDown: mouseDown, onMouseOut: mouseOut, onMouseMove: mouseMove }, ControlComponent)));
    }
    let left = startX || 0;
    let width = 0;
    if (startX !== undefined && currentX !== undefined) {
        left = currentX < startX ? currentX : startX;
        width = currentX - startX;
    }
    // calculate the start and end bp of drag
    let leftBpOffset;
    let rightBpOffset;
    if (startX) {
        leftBpOffset = overview.pxToBp(startX - cytobandOffset);
        rightBpOffset = overview.pxToBp(startX + width - cytobandOffset);
        if (currentX && currentX < startX) {
            ;
            [leftBpOffset, rightBpOffset] = [rightBpOffset, leftBpOffset];
        }
    }
    return (react_1.default.createElement("div", { style: { position: 'relative' } },
        rubberBandRef.current ? (react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement(material_1.Popover, { className: classes.popover, classes: {
                    paper: classes.paper,
                }, open: true, anchorEl: rubberBandRef.current, anchorOrigin: {
                    vertical: 'top',
                    horizontal: 'left',
                }, transformOrigin: {
                    vertical: 'bottom',
                    horizontal: 'right',
                }, keepMounted: true, disableRestoreFocus: true },
                react_1.default.createElement(material_1.Typography, null, leftBpOffset ? (0, util_1.stringify)(leftBpOffset) : '')),
            react_1.default.createElement(material_1.Popover, { className: classes.popover, classes: {
                    paper: classes.paper,
                }, open: true, anchorEl: rubberBandRef.current, anchorOrigin: {
                    vertical: 'top',
                    horizontal: 'right',
                }, transformOrigin: {
                    vertical: 'bottom',
                    horizontal: 'left',
                }, keepMounted: true, disableRestoreFocus: true },
                react_1.default.createElement(material_1.Typography, null, rightBpOffset ? (0, util_1.stringify)(rightBpOffset) : '')))) : null,
        react_1.default.createElement("div", { ref: rubberBandRef, className: classes.rubberBand, style: {
                left,
                width: Math.abs(width),
                height: __1.HEADER_OVERVIEW_HEIGHT,
            } }),
        react_1.default.createElement("div", { "data-testid": "rubberBand_controls", className: classes.rubberBandControl, role: "presentation", ref: controlsRef, onMouseDown: mouseDown, onMouseOut: mouseOut, onMouseMove: mouseMove }, ControlComponent)));
}
exports.default = (0, mobx_react_1.observer)(OverviewRubberBand);
