// @jbrowse/plugin-sequence/dist/BgzipFastaAdapter/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const configuration_1 = require("@jbrowse/core/configuration");
/**
 * #config BgzipFastaAdapter
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const BgzipFastaAdapter = (0, configuration_1.ConfigurationSchema)('BgzipFastaAdapter', {
    /**
     * #slot
     */
    fastaLocation: {
        type: 'fileLocation',
        defaultValue: { uri: '/path/to/seq.fa.gz', locationType: 'UriLocation' },
    },
    /**
     * #slot
     */
    faiLocation: {
        type: 'fileLocation',
        defaultValue: {
            uri: '/path/to/seq.fa.gz.fai',
            locationType: 'UriLocation',
        },
    },
    /**
     * #slot
     */
    metadataLocation: {
        description: 'Optional metadata file',
        type: 'fileLocation',
        defaultValue: {
            uri: '/path/to/fa.metadata.yaml',
            locationType: 'UriLocation',
        },
    },
    /**
     * #slot
     */
    gziLocation: {
        type: 'fileLocation',
        defaultValue: {
            uri: '/path/to/seq.fa.gz.gzi',
            locationType: 'UriLocation',
        },
    },
}, { explicitlyTyped: true });
exports.default = BgzipFastaAdapter;
