// @jbrowse/plugin-alignments/dist/LinearPileupDisplay/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.linearPileupDisplayConfigSchemaFactory = exports.linearPileupDisplayStateModelFactory = void 0;
const configSchema_1 = __importDefault(require("./configSchema"));
exports.linearPileupDisplayConfigSchemaFactory = configSchema_1.default;
const model_1 = __importDefault(require("./model"));
exports.linearPileupDisplayStateModelFactory = model_1.default;
const DisplayType_1 = __importDefault(require("@jbrowse/core/pluggableElementTypes/DisplayType"));
const plugin_linear_genome_view_1 = require("@jbrowse/plugin-linear-genome-view");
function register(pluginManager) {
    pluginManager.addDisplayType(() => {
        const configSchema = (0, configSchema_1.default)(pluginManager);
        return new DisplayType_1.default({
            name: 'LinearPileupDisplay',
            configSchema,
            stateModel: (0, model_1.default)(configSchema),
            trackType: 'AlignmentsTrack',
            viewType: 'LinearGenomeView',
            ReactComponent: plugin_linear_genome_view_1.BaseLinearDisplayComponent,
        });
    });
}
exports.default = register;
