// @jbrowse/react-linear-genome-view/esm/createViewState.js
import { autorun } from 'mobx';
import { onPatch } from 'mobx-state-tree';
import createModel from './createModel';
export default function createViewState(opts) {
    const { assembly, tracks, internetAccounts, configuration, aggregateTextSearchAdapters, plugins, location, onChange, disableAddTracks = false, makeWorkerInstance, } = opts;
    const { model, pluginManager } = createModel(plugins || [], makeWorkerInstance);
    let { defaultSession } = opts;
    if (!defaultSession) {
        defaultSession = {
            name: 'this session',
            view: {
                id: 'linearGenomeView',
                type: 'LinearGenomeView',
            },
        };
    }
    const stateTree = model.create({
        config: {
            configuration,
            assembly,
            tracks,
            internetAccounts,
            aggregateTextSearchAdapters,
        },
        disableAddTracks,
        session: defaultSession,
    }, { pluginManager });
    stateTree.config.internetAccounts.forEach(account => {
        const internetAccountType = pluginManager.getInternetAccountType(account.type);
        if (!internetAccountType) {
            throw new Error(`unknown internet account type ${account.type}`);
        }
        stateTree.addInternetAccount({
            type: account.type,
            configuration: account,
        });
    });
    pluginManager.setRootModel(stateTree);
    pluginManager.configure();
    if (location) {
        autorun(async (reaction) => {
            const { session } = stateTree;
            if (!session.view.initialized) {
                return;
            }
            if (typeof location === 'string') {
                session.view.navToLocString(location, assembly.name);
            }
            else {
                session.view.navTo(location);
            }
            reaction.dispose();
        });
    }
    if (onChange) {
        onPatch(stateTree, onChange);
    }
    return stateTree;
}
