// @jbrowse/plugin-data-management/dist/ucsc-trackhub/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const models_1 = require("@jbrowse/core/pluggableElementTypes/models");
const configuration_1 = require("@jbrowse/core/configuration");
/**
 * #config UCSCTrackHubConnection
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const UCSCTrackHubConnection = (0, configuration_1.ConfigurationSchema)('UCSCTrackHubConnection', {
    /**
     * #slot
     */
    hubTxtLocation: {
        type: 'fileLocation',
        defaultValue: {
            uri: 'http://mysite.com/path/to/hub.txt',
            locationType: 'UriLocation',
        },
        description: 'location of the hub file (usually called hub.txt)',
    },
    /**
     * #slot
     */
    assemblyNames: {
        type: 'stringArray',
        defaultValue: [],
        description: 'optional list of genomes to import from this track hub, if empty all genomes will be imported',
    },
}, {
    /**
     * #baseConfiguration
     */
    baseConfiguration: models_1.baseConnectionConfig,
});
exports.default = UCSCTrackHubConnection;
