// @jbrowse/plugin-config/dist/FromConfigAdapter/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sequenceConfigSchema = exports.regionsConfigSchema = exports.configSchema = exports.SequenceAdapterClass = exports.RegionsAdapterClass = exports.AdapterClass = void 0;
var FromConfigAdapter_1 = require("./FromConfigAdapter");
Object.defineProperty(exports, "AdapterClass", { enumerable: true, get: function () { return __importDefault(FromConfigAdapter_1).default; } });
var FromConfigRegionsAdapter_1 = require("./FromConfigRegionsAdapter");
Object.defineProperty(exports, "RegionsAdapterClass", { enumerable: true, get: function () { return __importDefault(FromConfigRegionsAdapter_1).default; } });
var FromConfigSequenceAdapter_1 = require("./FromConfigSequenceAdapter");
Object.defineProperty(exports, "SequenceAdapterClass", { enumerable: true, get: function () { return __importDefault(FromConfigSequenceAdapter_1).default; } });
var fromConfig_1 = require("./fromConfig");
Object.defineProperty(exports, "configSchema", { enumerable: true, get: function () { return __importDefault(fromConfig_1).default; } });
var fromConfig_2 = require("./fromConfig");
Object.defineProperty(exports, "regionsConfigSchema", { enumerable: true, get: function () { return __importDefault(fromConfig_2).default; } });
var fromConfigSequence_1 = require("./fromConfigSequence");
Object.defineProperty(exports, "sequenceConfigSchema", { enumerable: true, get: function () { return __importDefault(fromConfigSequence_1).default; } });
