// @jbrowse/plugin-sequence/dist/SequenceSearchAdapter/configSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const configuration_1 = require("@jbrowse/core/configuration");
/**
 * #config SequenceSearchAdapter
 */
function x() { } // eslint-disable-line @typescript-eslint/no-unused-vars
const configSchema = (0, configuration_1.ConfigurationSchema)('SequenceSearchAdapter', {
    /**
     * #slot
     */
    search: {
        type: 'string',
        defaultValue: '',
        description: 'Search string or regex to search for',
    },
    /**
     * #slot
     */
    sequenceAdapter: {
        type: 'frozen',
        defaultValue: null,
    },
    /**
     * #slot
     */
    searchForward: {
        type: 'boolean',
        defaultValue: true,
    },
    /**
     * #slot
     */
    searchReverse: {
        type: 'boolean',
        defaultValue: true,
    },
    /**
     * #slot
     */
    caseInsensitive: {
        type: 'boolean',
        defaultValue: true,
    },
}, { explicitlyTyped: true });
exports.default = configSchema;
