// @gmod/vcf/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseBreakend = void 0;
var parse_1 = __importDefault(require("./parse"));
function parseBreakend(breakendString) {
    var tokens = breakendString.split(/[[\]]/);
    if (tokens.length > 1) {
        var MateDirection = breakendString.includes('[') ? 'right' : 'left';
        var Join = void 0;
        var Replacement = void 0;
        var MatePosition = void 0;
        for (var i = 0; i < tokens.length; i += 1) {
            var tok = tokens[i];
            if (tok) {
                if (tok.includes(':')) {
                    // this is the remote location
                    MatePosition = tok;
                    Join = Replacement ? 'right' : 'left';
                }
                else {
                    // this is the local alteration
                    Replacement = tok;
                }
            }
        }
        if (!(MatePosition && Join && Replacement)) {
            throw new Error("Invalid breakend: ".concat(breakendString));
        }
        return { MatePosition: MatePosition, Join: Join, Replacement: Replacement, MateDirection: MateDirection };
    }
    else {
        if (breakendString.startsWith('.')) {
            return {
                Join: 'left',
                SingleBreakend: true,
                Replacement: breakendString.slice(1),
            };
        }
        else if (breakendString.endsWith('.')) {
            return {
                Join: 'right',
                SingleBreakend: true,
                Replacement: breakendString.slice(0, breakendString.length - 1),
            };
        }
        else if (breakendString[0] === '<') {
            var res = breakendString.match('<(.*)>(.*)');
            if (!res) {
                throw new Error("failed to parse ".concat(breakendString));
            }
            var Replacement = res === null || res === void 0 ? void 0 : res[2];
            return Replacement
                ? {
                    Join: 'left',
                    Replacement: Replacement,
                    MateDirection: 'right',
                    MatePosition: "<".concat(res === null || res === void 0 ? void 0 : res[1], ">:1"),
                }
                : undefined;
        }
        else if (breakendString.includes('<')) {
            var res = breakendString.match('(.*)<(.*)>');
            if (!res) {
                throw new Error("failed to parse ".concat(breakendString));
            }
            var Replacement = res === null || res === void 0 ? void 0 : res[1];
            return Replacement
                ? {
                    Join: 'right',
                    Replacement: Replacement,
                    MateDirection: 'right',
                    MatePosition: "<".concat(res === null || res === void 0 ? void 0 : res[2], ">:1"),
                }
                : undefined;
        }
    }
    return undefined;
}
exports.parseBreakend = parseBreakend;
exports.default = parse_1.default;
