// @jbrowse/plugin-sequence/dist/ChromSizesAdapter/ChromSizesAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const BaseAdapter_1 = require("@jbrowse/core/data_adapters/BaseAdapter");
const io_1 = require("@jbrowse/core/util/io");
class default_1 extends BaseAdapter_1.BaseAdapter {
    async setupPre() {
        const pm = this.pluginManager;
        const file = (0, io_1.openLocation)(this.getConf('chromSizesLocation'), pm);
        const data = await file.readFile('utf8');
        return Object.fromEntries(data
            .split(/\n|\r\n|\r/)
            .map(f => f.trim())
            .filter(f => !!f)
            .map((line) => {
            const [name, length] = line.split('\t');
            return [name, +length];
        }));
    }
    async setup() {
        if (!this.setupP) {
            this.setupP = this.setupPre().catch(e => {
                this.setupP = undefined;
                throw e;
            });
        }
        return this.setupP;
    }
    async getRegions() {
        const refSeqs = await this.setup();
        return Object.keys(refSeqs).map(refName => ({
            refName,
            start: 0,
            end: refSeqs[refName],
        }));
    }
    getHeader() {
        return {};
    }
    freeResources( /* { region } */) { }
}
exports.default = default_1;
