// @gmod/nclist/dist/lazy_array.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __asyncDelegator = (this && this.__asyncDelegator) || function (o) {
    var i, p;
    return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
    function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
};
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
    function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
//@ts-nocheck
var url_1 = __importDefault(require("url"));
var quick_lru_1 = __importDefault(require("quick-lru"));
var abortable_promise_cache_1 = __importDefault(require("abortable-promise-cache"));
var util_1 = require("./util");
/**
 * For a JSON array that gets too large to load in one go, this class
 * helps break it up into chunks and provides an
 * async API for using the information in the array.
 */
var LazyArray = /** @class */ (function () {
    function LazyArray(_a, baseUrl) {
        var urlTemplate = _a.urlTemplate, chunkSize = _a.chunkSize, length = _a.length, _b = _a.cacheSize, cacheSize = _b === void 0 ? 100 : _b, readFile = _a.readFile;
        this.urlTemplate = urlTemplate;
        this.chunkSize = chunkSize;
        this.length = length;
        this.baseUrl = baseUrl === undefined ? '' : baseUrl;
        this.readFile = readFile;
        if (!readFile) {
            throw new Error('must provide readFile callback');
        }
        this.chunkCache = new abortable_promise_cache_1.default({
            cache: new quick_lru_1.default({ maxSize: cacheSize }),
            fill: this.getChunk.bind(this),
        });
    }
    /**
     * call the callback on one element of the array
     * @param i index
     * @param callback callback, gets called with (i, value, param)
     * @param param (optional) callback will get this as its last parameter
     */
    LazyArray.prototype.index = function (i, callback, param) {
        this.range(i, i, callback, undefined, param);
    };
    /**
     * async generator for the elements in the range [start,end]
     *
     * @param start index of first element to call the callback on
     * @param end index of last element to call the callback on
     */
    LazyArray.prototype.range = function (start, end) {
        return __asyncGenerator(this, arguments, function range_1() {
            var firstChunk, lastChunk, chunkreadFiles, chunk, i, _a, chunkNumber, chunkData;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        start = Math.max(0, start);
                        end = Math.min(end, this.length - 1);
                        firstChunk = Math.floor(start / this.chunkSize);
                        lastChunk = Math.floor(end / this.chunkSize);
                        chunkreadFiles = [];
                        for (chunk = firstChunk; chunk <= lastChunk; chunk += 1) {
                            chunkreadFiles.push(this.chunkCache.get(chunk, chunk));
                        }
                        i = 0;
                        _b.label = 1;
                    case 1:
                        if (!(i < chunkreadFiles.length)) return [3 /*break*/, 6];
                        return [4 /*yield*/, __await(chunkreadFiles[i])];
                    case 2:
                        _a = _b.sent(), chunkNumber = _a[0], chunkData = _a[1];
                        return [5 /*yield**/, __values(__asyncDelegator(__asyncValues(this.filterChunkData(start, end, chunkNumber, chunkData))))];
                    case 3: return [4 /*yield*/, __await.apply(void 0, [_b.sent()])];
                    case 4:
                        _b.sent();
                        _b.label = 5;
                    case 5:
                        i += 1;
                        return [3 /*break*/, 1];
                    case 6: return [2 /*return*/];
                }
            });
        });
    };
    LazyArray.prototype.getChunk = function (chunkNumber) {
        return __awaiter(this, void 0, void 0, function () {
            var url, data;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        url = this.urlTemplate.replace(/\{Chunk\}/gi, chunkNumber);
                        if (this.baseUrl) {
                            url = url_1.default.resolve(this.baseUrl, url);
                        }
                        return [4 /*yield*/, (0, util_1.readJSON)(url, this.readFile)];
                    case 1:
                        data = _a.sent();
                        return [2 /*return*/, [chunkNumber, data]];
                }
            });
        });
    };
    LazyArray.prototype.filterChunkData = function (queryStart, queryEnd, chunkNumber, chunkData) {
        var firstIndex, chunkStart, chunkEnd, i;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    firstIndex = chunkNumber * this.chunkSize;
                    chunkStart = Math.max(0, queryStart - firstIndex);
                    chunkEnd = Math.min(queryEnd - firstIndex, this.chunkSize - 1);
                    i = chunkStart;
                    _a.label = 1;
                case 1:
                    if (!(i <= chunkEnd)) return [3 /*break*/, 4];
                    return [4 /*yield*/, [i + firstIndex, chunkData[i]]];
                case 2:
                    _a.sent();
                    _a.label = 3;
                case 3:
                    i += 1;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/];
            }
        });
    };
    return LazyArray;
}());
exports.default = LazyArray;
