// @jbrowse/react-linear-genome-view/dist/JBrowseLinearGenomeView/AboutDialog.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const version_1 = require("../version");
const material_1 = require("@mui/material");
const Close_1 = __importDefault(require("@mui/icons-material/Close"));
const mui_1 = require("tss-react/mui");
const useStyles = (0, mui_1.makeStyles)()(theme => ({
    closeButton: {
        position: 'absolute',
        right: theme.spacing(1),
        top: theme.spacing(1),
        color: theme.palette.grey[500],
    },
    content: {
        minWidth: 800,
    },
}));
function AboutDialog({ open, onClose, }) {
    const { classes } = useStyles();
    return (react_1.default.createElement(material_1.Dialog, { open: open, onClose: onClose, maxWidth: "xl" },
        react_1.default.createElement(material_1.DialogTitle, null,
            "JBrowse v",
            version_1.version,
            react_1.default.createElement(material_1.IconButton, { className: classes.closeButton, onClick: () => onClose(), size: "large" },
                react_1.default.createElement(Close_1.default, null))),
        react_1.default.createElement(material_1.DialogContent, { className: classes.content },
            react_1.default.createElement(material_1.Typography, null, "JBrowse is a GMOD project \u00A9 2019-2021, The Evolutionary Software Foundation"),
            react_1.default.createElement(material_1.Typography, null, "Here are some resources and documentation. Please report the version number above when asking questions. Thanks!"),
            react_1.default.createElement("ul", null,
                react_1.default.createElement("li", null,
                    react_1.default.createElement(material_1.Link, { href: "https://github.com/GMOD/jbrowse-components/discussions", target: "_blank", rel: "noopener noreferrer" }, "Question & answer forum")),
                react_1.default.createElement("li", null,
                    react_1.default.createElement(material_1.Link, { href: "https://github.com/GMOD/jbrowse-components/issues/new/choose", target: "_blank", rel: "noopener noreferrer" }, "Report a bug")),
                react_1.default.createElement("li", null,
                    react_1.default.createElement(material_1.Link, { href: "https://jbrowse.org/jb2/docs/user_guide", target: "_blank", rel: "noopener noreferrer" }, "User guide")),
                react_1.default.createElement("li", null,
                    react_1.default.createElement(material_1.Link, { href: "https://jbrowse.org/jb2/docs/", target: "_blank", rel: "noopener noreferrer" }, "Documentation"))))));
}
exports.default = AboutDialog;
